oak_functions/lookup/src/lib.rs

```diff
@@ -115,30 +146,57 @@ pub type Data = HashMap<Vec<u8>, Vec<u8>>;
 /// In the future we may replace both the mutex and the hash map with something like RCU.
 pub struct LookupDataManager<L: OakLogger + Clone> {
     data: Spinlock<Arc<Data>>,
+    // Behind a lock, because we have multiple references to LookupDataManager and need to mutate
+    // data builder.
+    data_builder: Spinlock<DataBuilder>,
     logger: L,
 }
 
+#[derive(Clone)]
+pub enum UpdateAction {
+    StartAndFinish,
+}
+
+pub enum UpdateStatus {
+    Started,
+    Finished,
+    Aborted,
+}
+
 impl<L> LookupDataManager<L>
 where
     L: OakLogger + Clone,
 {
     /// Creates a new instance with empty backing data.
     pub fn new_empty(logger: L) -> Self {
         Self {
-            data: Spinlock::new(Arc::new(HashMap::new())),
+            data: Spinlock::new(Arc::new(Data::new())),
+            data_builder: Spinlock::new(DataBuilder::new()),
             logger,
         }
     }
 
     /// Creates an instance of LookupData populated with the given entries.
-    pub fn for_test(entries: Data, logger: L) -> Self {
-        let data = Spinlock::new(Arc::new(entries));
-        Self { data, logger }
+    pub fn for_test(data: Data, logger: L) -> Self {
+        let test_manager = Self::new_empty(logger);
+        *test_manager.data.lock() = Arc::new(data);
+        test_manager
     }
 
     /// Updates the backing data that will be used by new `LookupData` instances.
-    pub fn update_data(&self, data: Data) {
-        *self.data.lock() = Arc::new(data);
+    pub fn update_data(&self, action: UpdateAction, new_data: Data) -> UpdateStatus {
```

Comment:
In terms of API design, even if the action is represented as an enum value in protobuf, I would expect there to be separate methods on this Rust struct related to updating the data (basically just calling the underlying extend function), and another that just makes it current (instead of doing two things in the same method based on the enum value).
Anyways I see the value of having this method match the proto API, so up to you.

Reply:
I would keep a single matching `update_lookup_data` in the LookupDataManager for the moment. 

It'll keep the update logic contained in one place.

I am happy to change later!